File: swift-bridge/Sources/CoreAnimationBridge/CAShapeLayer.swift
import Foundation
import QuartzCore

private func caShapeFillRule(_ raw: Int32) -> CAShapeLayerFillRule {
    raw == 1 ? .evenOdd : .nonZero
}

private func caShapeFillRuleRaw(_ value: CAShapeLayerFillRule) -> Int32 {
    value == .evenOdd ? 1 : 0
}

@_cdecl("ca_shape_layer_get_fill_rule")
public func ca_shape_layer_get_fill_rule(_ handle: UnsafeMutableRawPointer?) -> Int32 {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return 0 }
    return caShapeFillRuleRaw(layer.fillRule)
}

@_cdecl("ca_shape_layer_set_fill_rule")
public func ca_shape_layer_set_fill_rule(_ handle: UnsafeMutableRawPointer?, _ value: Int32) {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return }
    layer.fillRule = caShapeFillRule(value)
}

@_cdecl("ca_shape_layer_get_stroke_start")
public func ca_shape_layer_get_stroke_start(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return 0 }
    return layer.strokeStart
}

@_cdecl("ca_shape_layer_set_stroke_start")
public func ca_shape_layer_set_stroke_start(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return }
    layer.strokeStart = value
}

@_cdecl("ca_shape_layer_get_stroke_end")
public func ca_shape_layer_get_stroke_end(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return 0 }
    return layer.strokeEnd
}

@_cdecl("ca_shape_layer_set_stroke_end")
public func ca_shape_layer_set_stroke_end(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return }
    layer.strokeEnd = value
}

@_cdecl("ca_shape_layer_get_line_dash_phase")
public func ca_shape_layer_get_line_dash_phase(_ handle: UnsafeMutableRawPointer?) -> Double {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return 0 }
    return layer.lineDashPhase
}

@_cdecl("ca_shape_layer_set_line_dash_phase")
public func ca_shape_layer_set_line_dash_phase(_ handle: UnsafeMutableRawPointer?, _ value: Double) {
    guard let layer: CAShapeLayer = caBorrow(handle) else { return }
    layer.lineDashPhase = value
}

File: src/ca_shape_layer.rs
use crate::layer::{LayerLike, ShapeLayer};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(i32)]
pub enum ShapeFillRule {
    NonZero = 0,
    EvenOdd = 1,
}

impl ShapeFillRule {
    const fn from_raw(value: i32) -> Self {
        match value {
            1 => Self::EvenOdd,
            _ => Self::NonZero,
        }
    }
}

impl ShapeLayer {
    #[must_use]
    pub fn fill_rule(&self) -> ShapeFillRule {
        ShapeFillRule::from_raw(unsafe {
            crate::ffi::ca_shape_layer_get_fill_rule(self.as_layer_ptr())
        })
    }

    pub fn set_fill_rule(&self, value: ShapeFillRule) {
        unsafe { crate::ffi::ca_shape_layer_set_fill_rule(self.as_layer_ptr(), value as i32) };
    }

    #[must_use]
    pub fn stroke_start(&self) -> f64 {
        unsafe { crate::ffi::ca_shape_layer_get_stroke_start(self.as_layer_ptr()) }
    }

    pub fn set_stroke_start(&self, value: f64) {
        unsafe { crate::ffi::ca_shape_layer_set_stroke_start(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn stroke_end(&self) -> f64 {
        unsafe { crate::ffi::ca_shape_layer_get_stroke_end(self.as_layer_ptr()) }
    }

    pub fn set_stroke_end(&self, value: f64) {
        unsafe { crate::ffi::ca_shape_layer_set_stroke_end(self.as_layer_ptr(), value) };
    }

    #[must_use]
    pub fn line_dash_phase(&self) -> f64 {
        unsafe { crate::ffi::ca_shape_layer_get_line_dash_phase(self.as_layer_ptr()) }
    }

    pub fn set_line_dash_phase(&self, value: f64) {
        unsafe { crate::ffi::ca_shape_layer_set_line_dash_phase(self.as_layer_ptr(), value) };
    }
}
